py/obj.h

```diff
@@ -184,13 +184,16 @@ static inline bool mp_obj_is_small_int(mp_const_obj_t o) {
 #define MP_OBJ_NEW_SMALL_INT(small_int) ((mp_obj_t)((((mp_uint_t)(small_int)) << 1) | 1))
 
 #if MICROPY_PY_BUILTINS_FLOAT
-#define MP_OBJ_NEW_CONST_FLOAT(f) MP_ROM_PTR((mp_obj_t)((((((uint64_t)f) & ~3) | 2) + 0x80800000) & 0xffffffff))
+#include <math.h>
+#define mp_nan_obj_bits (0x7fc00000)
+#define MP_OBJ_NEW_FLOAT(f) (mp_obj_t)((((((uint64_t)f) & ~3) | 2) + 0x80800000) & 0xffffffff)
+#define MP_OBJ_NEW_CONST_FLOAT(f) MP_ROM_PTR(MP_OBJ_NEW_FLOAT(f))
```

Comment:
I can understand why you've done things this way, but IMO it's a little complex and this code here is already complex enough.

So I suggest just using the known definition of `MP_ROM_PTR` for repr C, which is just the identity, and doing:
```c
#define MP_OBJ_NEW_CONST_FLOAT(f) ((mp_obj_t)((((((uint64_t)f) & ~3) | 2) + 0x80800000) & 0xffffffff))
```
ie drop `MP_ROM_PTR`.  Then `mp_const_float_nan` can be used directly below.

I think that will lead to the most readable code.

You could add a small comment above the definition of `MP_OBJ_NEW_CONST_FLOAT` saying something like "should be a MP_ROM_PTR but that macro isn't available yet".